<a href="https://colab.research.google.com/github/yuniastariiy/IBM-Data-Analyst---Phyton/blob/main/Activity_Course_2_TikTok_project_lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **TikTok Project**
**Course 2 - Get Started with Python**

Welcome to the TikTok Project!

You have just started as a data professional at TikTok.

The team is still in the early stages of the project. You have received notice that TikTok's leadership team has approved the project proposal. To gain clear insights to prepare for a claims classification model, TikTok's provided data must be examined to begin the process of exploratory data analysis (EDA).

A notebook was structured and prepared to help you in this project. Please complete the following questions.

# **Course 2 End-of-course project: Inspect and analyze data**

In this activity, you will examine data provided and prepare it for analysis.
<br/>

**The purpose** of this project is to investigate and understand the data provided. This activity will:

1.   Acquaint you with the data

2.   Compile summary information about the data

3.   Begin the process of EDA and reveal insights contained in the data

4.   Prepare you for more in-depth EDA, hypothesis testing, and statistical analysis

**The goal** is to construct a dataframe in Python, perform a cursory inspection of the provided dataset, and inform TikTok data team members of your findings.
<br/>
*This activity has three parts:*

**Part 1:** Understand the situation
* How can you best prepare to understand and organize the provided TikTok information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning and future exploratory data analysis (EDA) and statistical activities

* Compile summary information about the data to inform next steps

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into variables

<br/>

To complete the activity, follow the instructions and answer the questions below. Then, you will us your responses to these questions and the questions included in the Course 2 PACE Strategy Document to create an executive summary.

Be sure to complete this activity before moving on to Course 3. You can assess your work by comparing the results to a completed exemplar after completing the end-of-course project.

# **Identify data types and compile summary information**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

# **PACE stages**

   *        [Plan](#scrollTo=psz51YkZVwtN&line=3&uniqifier=1)
   *        [Analyze](#scrollTo=mA7Mz_SnI8km&line=4&uniqifier=1)
   *        [Construct](#scrollTo=Lca9c8XON8lc&line=2&uniqifier=1)
   *        [Execute](#scrollTo=401PgchTPr4E&line=2&uniqifier=1)

## **PACE: Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:



### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided information?


*Begin by exploring your dataset and consider reviewing the Data Dictionary.*


To best prepare to understand and organize the provided TikTok dataset for the claims classification model, the following steps should be taken:

1. **Explore the Dataset**  
   Start by loading and visually inspecting the dataset using Python libraries such as `pandas`. This includes:
   - Checking the shape of the dataset (number of rows and columns)
   - Displaying the first few rows to understand the structure
   - Reviewing column names and data types
   - Identifying missing values or unusual entries

2. **Review the Data Dictionary**  
   The data dictionary provides definitions and descriptions for each variable in the dataset. This helps:
   - Understand what each column represents
   - Determine which features may be useful for analysis or modeling
   - Identify potential data cleaning or preprocessing steps

By combining these steps, we can develop a solid understanding of the dataset’s structure and content, which will guide future stages of exploratory data analysis (EDA) and model preparation.


## **PACE: Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Imports and data loading**

Start by importing the packages that you will need to load and explore the dataset. Make sure to use the following import statements:
*   `import pandas as pd`

*   `import numpy as np`


In [ ]:
# Import packages
import pandas as pd
import numpy as np

Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

### **Task 2b. Understand the data - Inspect the data**

View and inspect summary information about the dataframe by **coding the following:**

1. `data.head(10)`
2. `data.info()`
3. `data.describe()`

*Consider the following questions:*

**Question 1:** When reviewing the first few rows of the dataframe, what do you observe about the data? What does each row represent?

**Question 2:** When reviewing the `data.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Question 3:** When reviewing the `data.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values? Does it seem that there are outlier values?

















In [ ]:
# Load dataset into dataframe
data = pd.read_csv("tiktok_dataset.csv")
data

,#,claim_status,video_id,video_duration_sec,video_transcription_text,verified_status,author_ban_status,video_view_count,video_like_count,video_share_count,video_download_count,video_comment_count
0,1,claim,7017666017,59,someone shared with me that drone deliveries a...,not verified,under review,343296.0,19425.0,241.0,1.0,0.0
1,2,claim,4014381136,32,someone shared with me that there are more mic...,not verified,active,140877.0,77355.0,19034.0,1161.0,684.0
2,3,claim,9859838091,31,someone shared with me that american industria...,not verified,active,902185.0,97690.0,2858.0,833.0,329.0
3,4,claim,1866847991,25,someone shared with me that the metro of st. p...,not verified,active,437506.0,239954.0,34812.0,1234.0,584.0
4,5,claim,7105231098,19,someone shared with me that the number of busi...,not verified,active,56167.0,34987.0,4110.0,547.0,152.0
...,...,...,...,...,...,...,...,...,...,...,...,...
19377,19378,NaN,7578226840,21,NaN,not verified,active,NaN,NaN,NaN,NaN,NaN
19378,19379,NaN,6079236179,53,NaN,not verified,active,NaN,NaN,NaN,NaN,NaN
19379,19380,NaN,2565539685,10,NaN,verified,under review,NaN,NaN,NaN,NaN,NaN
19380,19381,NaN,2969178540,24,NaN,not verified,active,NaN,NaN,NaN,NaN,NaN


In [ ]:
# Display and examine the first ten rows of the dataframe
data.head(10)


,#,claim_status,video_id,video_duration_sec,video_transcription_text,verified_status,author_ban_status,video_view_count,video_like_count,video_share_count,video_download_count,video_comment_count
0,1,claim,7017666017,59,someone shared with me that drone deliveries a...,not verified,under review,343296.0,19425.0,241.0,1.0,0.0
1,2,claim,4014381136,32,someone shared with me that there are more mic...,not verified,active,140877.0,77355.0,19034.0,1161.0,684.0
2,3,claim,9859838091,31,someone shared with me that american industria...,not verified,active,902185.0,97690.0,2858.0,833.0,329.0
3,4,claim,1866847991,25,someone shared with me that the metro of st. p...,not verified,active,437506.0,239954.0,34812.0,1234.0,584.0
4,5,claim,7105231098,19,someone shared with me that the number of busi...,not verified,active,56167.0,34987.0,4110.0,547.0,152.0
5,6,claim,8972200955,35,someone shared with me that gross domestic pro...,not verified,under review,336647.0,175546.0,62303.0,4293.0,1857.0
6,7,claim,4958886992,16,someone shared with me that elvis presley has ...,not verified,active,750345.0,486192.0,193911.0,8616.0,5446.0
7,8,claim,2270982263,41,someone shared with me that the best selling s...,not verified,active,547532.0,1072.0,50.0,22.0,11.0
8,9,claim,5235769692,50,someone shared with me that about half of the ...,not verified,active,24819.0,10160.0,1050.0,53.0,27.0
9,10,claim,4660861094,45,someone shared with me that it would take a 50...,verified,active,931587.0,171051.0,67739.0,4104.0,2540.0


In [ ]:
# Get summary info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19382 entries, 0 to 19381
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   #                         19382 non-null  int64  
 1   claim_status              19084 non-null  object 
 2   video_id                  19382 non-null  int64  
 3   video_duration_sec        19382 non-null  int64  
 4   video_transcription_text  19084 non-null  object 
 5   verified_status           19382 non-null  object 
 6   author_ban_status         19382 non-null  object 
 7   video_view_count          19084 non-null  float64
 8   video_like_count          19084 non-null  float64
 9   video_share_count         19084 non-null  float64
 10  video_download_count      19084 non-null  float64
 11  video_comment_count       19084 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 1.8+ MB


In [ ]:
# Get summary statistics
data.describe()

,#,video_id,video_duration_sec,video_view_count,video_like_count,video_share_count,video_download_count,video_comment_count
count,19382.000000,1.938200e+04,19382.000000,19084.000000,19084.000000,19084.000000,19084.000000,19084.000000
mean,9691.500000,5.627454e+09,32.421732,254708.558688,84304.636030,16735.248323,1049.429627,349.312146
std,5595.245794,2.536440e+09,16.229967,322893.280814,133420.546814,32036.174350,2004.299894,799.638865
min,1.000000,1.234959e+09,5.000000,20.000000,0.000000,0.000000,0.000000,0.000000
25%,4846.250000,3.430417e+09,18.000000,4942.500000,810.750000,115.000000,7.000000,1.000000
50%,9691.500000,5.618664e+09,32.000000,9954.500000,3403.500000,717.000000,46.000000,9.000000
75%,14536.750000,7.843960e+09,47.000000,504327.000000,125020.000000,18222.000000,1156.250000,292.000000
max,19382.000000,9.999873e+09,60.000000,999817.000000,657830.000000,256130.000000,14994.000000,9599.000000


#### **Question 1: When reviewing the first few rows of the dataframe, what do you observe about the data? What does each row represent?**

Each row in the dataset represents a single TikTok video entry that is part of a claim classification project. The rows contain various details related to the video, such as its ID, duration, transcription, claim status, whether the user is verified or banned, and engagement metrics like views, likes, shares, downloads, and comments. This indicates that each row captures the metadata and engagement details of one unique video.

---

#### **Question 2: When reviewing the `data.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?**

- The dataset consists of 12 columns and 19,382 rows.
- Some columns have missing values, particularly `claim_status`, `video_transcription_text`, and all engagement metric columns (from `video_view_count` to `video_comment_count`) — each missing around 298 values.
- Data types include `int64`, `float64`, and `object`, meaning there is a mix of numeric and categorical/text data.
- The `video_id` column is numeric but likely represents unique identifiers better stored as strings.
- No date/time columns are present.
- The presence of missing values in key fields may require data cleaning or imputation in further analysis.

---

#### **Question 3: When reviewing the `data.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values? Does it seem that there are outlier values?**

- The numerical data shows signs of **skewed distributions** with means significantly higher than medians.
- Several engagement metrics, such as `video_like_count` and `video_comment_count`, display **potential outliers**—with very high max values but relatively low 75th percentiles.
- For instance, `video_view_count` ranges from 5 to almost 1 million views, but the 75th percentile is only 47.
- Some metrics have a **minimum of 0**, indicating that some videos had no engagement at all, which may be valid but worth further examination.
- Overall, the dataset likely includes **viral content outliers**, which is common on platforms like TikTok and should be considered during further statistical or machine learning analysis.


### **Task 2c. Understand the data - Investigate the variables**

In this phase, you will begin to investigate the variables more closely to better understand them.

You know from the project proposal that the ultimate objective is to use machine learning to classify videos as either claims or opinions. A good first step towards understanding the data might therefore be examining the `claim_status` variable. Begin by determining how many videos there are for each different claim status.

In [ ]:
# Check unique values and their counts for the 'claim_status' variable
data['claim_status'].value_counts(dropna=False)

claim      9608
opinion    9476
NaN         298
Name: claim_status, dtype: int64

**Question:** What do you notice about the values shown?

The `claim_status` variable contains three types of values:

- `claim`: 9,608 entries
- `opinion`: 9,476 entries
- `NaN`: 298 entries (missing values)

This indicates that the dataset is relatively balanced between claims and opinions, which is favorable for training a machine learning model. However, we should take note of the 298 missing values in `claim_status`, as these could impact the analysis and model performance. Handling these missing values appropriately (e.g., removing or imputing them) will be important in the data preprocessing phase.


Next, examine the engagement trends associated with each different claim status.

Start by using Boolean masking to filter the data according to claim status, then calculate the mean and median view counts for each claim status.

In [ ]:
claim_videos = data[data['claim_status'] == 'claim']
opinion_videos = data[data['claim_status'] == 'opinion']

In [ ]:
# Mean and median view count for 'claim' status
claim_mean_views = claim_videos['video_view_count'].mean()
claim_median_views = claim_videos['video_view_count'].median()

In [ ]:
print(f"Claim - Mean Views: {claim_mean_views:.0f}")
print(f"Claim - Median Views: {claim_median_views:.0f}")

Claim - Mean Views: 501029
Claim - Median Views: 501555


In [ ]:
# Mean and median view count for 'opinion' status
opinion_mean_views = opinion_videos['video_view_count'].mean()
opinion_median_views = opinion_videos['video_view_count'].median()

In [ ]:

print(f"Opinion - Mean Views: {opinion_mean_views:.0f}")
print(f"Opinion - Median Views: {opinion_median_views:.0f}")

Opinion - Mean Views: 4956
Opinion - Median Views: 4953


**Question:** What do you notice about the mean and median within each claim category?

**Answer:**  
There is a significant difference in both the mean and median view counts between the two claim categories. Videos labeled as **"claim"** have a much higher average and median view count (mean ≈ 501,029; median ≈ 501,555) compared to videos labeled as **"opinion"** (mean ≈ 4,956; median ≈ 4,953). This suggests that claim videos tend to receive considerably more views than opinion videos. The close proximity between the mean and median within each category also indicates that the distribution of view counts is relatively symmetrical and not heavily skewed within each group.


Now, examine trends associated with the ban status of the author.

Use `groupby()` to calculate how many videos there are for each combination of categories of claim status and author ban status.

In [ ]:
# Group by claim_status and author_ban_status, then count
claim_ban_counts = data.groupby(['claim_status', 'author_ban_status']).size().reset_index(name='video_count')

In [ ]:
# Display result
claim_ban_counts

,claim_status,author_ban_status,video_count
0,claim,active,6566
1,claim,banned,1439
2,claim,under review,1603
3,opinion,active,8817
4,opinion,banned,196
5,opinion,under review,463


**Question:** What do you notice about the number of claims videos with banned authors? Why might this relationship occur?

**Answer:**  
There are **1,439** claim videos associated with **banned authors**, which is notably higher than the **196** opinion videos associated with banned authors. This suggests that **claim videos are more frequently associated with banned accounts** compared to opinion videos.

This relationship may occur because **claims**—especially those that are misleading, false, or violate community guidelines—are **more likely to be flagged or reported**, leading to **account bans**. In contrast, **opinions** are typically subjective and less likely to breach platform policies, resulting in fewer bans. Therefore, the higher number of bans among authors of claim videos could reflect TikTok’s enforcement of content moderation policies against misinformation or harmful content.


Continue investigating engagement levels, now focusing on `author_ban_status`.

Calculate the median video share count of each author ban status.

In [ ]:
# Group by author_ban_status and calculate median of video_share_count
median_share_counts = data.groupby("author_ban_status")["video_share_count"].median()
print(median_share_counts)


author_ban_status
active            437.0
banned          14468.0
under review     9444.0
Name: video_share_count, dtype: float64


**Question: What do you notice about the share count of banned authors, compared to that of active authors? Explore this in more depth.**

Videos posted by **banned authors** have a **much higher median share count** (14,468) compared to those posted by **active authors** (437). This suggests that banned authors may have posted content that was more provocative, controversial, or engaging, leading to significantly more shares.

The high share counts could indicate that their videos went viral before being flagged or reviewed. In contrast, active authors have lower median shares, which may reflect more typical or less controversial engagement.

This pattern might imply a relationship between high engagement and content violations, where viral videos from banned accounts may have attracted scrutiny due to potential misinformation or harmful content.


Use `groupby()` to group the data by `author_ban_status`, then use `agg()` to get the count, mean, and median of each of the following columns:
* `video_view_count`
* `video_like_count`
* `video_share_count`

Remember, the argument for the `agg()` function is a dictionary whose keys are columns. The values for each column are a list of the calculations you want to perform.

In [ ]:
# Group data by author_ban_status and aggregate count, mean, and median for specified columns
ban_status_summary = data.groupby("author_ban_status")[["video_view_count", "video_like_count", "video_share_count"]].agg(["count", "mean", "median"])
ban_status_summary


video_view_count                          video_like_count  \
                             count           mean    median            count   
author_ban_status                                                              
active                       15383  215927.039524    8616.0            15383   
banned                        1635  445845.439144  448201.0             1635   
under review                  2066  392204.836399  365245.5             2066   

                                           video_share_count                \
                            mean    median             count          mean   
author_ban_status                                                            
active              71036.533836    2222.0             15383  14111.466164   
banned             153017.236697  105573.0              1635  29998.942508   
under review       128718.050339   71204.5              2066  25774.696999   

                            
                    median  
author_ban_status           
active               437.0  
banned             14468.0  
under review        9444.0

**Question:** What do you notice about the number of views, likes, and shares for banned authors compared to active authors?

**Answer:**
Videos from **banned authors** tend to receive significantly **higher engagement** than those from **active authors**:

- **Views**: Banned authors have a **mean view count of ~445,845**, compared to **~215,927** for active authors.
- **Likes**: Banned authors average **~153,017 likes**, whereas active authors have only **~71,037 likes** on average.
- **Shares**: Banned authors’ videos have a **median share count of 14,468**, much higher than the **437** for active authors.

This suggests that content from banned authors might be more **viral or controversial**, leading to greater engagement. However, it may also indicate that high engagement could correlate with content that violates community guidelines, resulting in bans.

Now, create three new columns to help better understand engagement rates:
* `likes_per_view`: represents the number of likes divided by the number of views for each video
* `comments_per_view`: represents the number of comments divided by the number of views for each video
* `shares_per_view`: represents the number of shares divided by the number of views for each video

In [ ]:
# Create new columns for engagement rates
data['likes_per_view'] = data['video_like_count'] / data['video_view_count']
data['comments_per_view'] = data['video_comment_count'] / data['video_view_count']
data['shares_per_view'] = data['video_share_count'] / data['video_view_count']

# Display the first few rows to confirm the new columns are added
data[['video_view_count', 'video_like_count', 'likes_per_view',
      'video_comment_count', 'comments_per_view',
      'video_share_count', 'shares_per_view']].head()


,video_view_count,video_like_count,likes_per_view,video_comment_count,comments_per_view,video_share_count,shares_per_view
0,343296.0,19425.0,0.056584,0.0,0.000000,241.0,0.000702
1,140877.0,77355.0,0.549096,684.0,0.004855,19034.0,0.135111
2,902185.0,97690.0,0.108282,329.0,0.000365,2858.0,0.003168
3,437506.0,239954.0,0.548459,584.0,0.001335,34812.0,0.079569
4,56167.0,34987.0,0.622910,152.0,0.002706,4110.0,0.073175


Use `groupby()` to compile the information in each of the three newly created columns for each combination of categories of claim status and author ban status, then use `agg()` to calculate the count, the mean, and the median of each group.

In [ ]:
# Group by claim_status and author_ban_status, then aggregate the new engagement rate columns
engagement_summary = data.groupby(['claim_status', 'author_ban_status'])[
    ['likes_per_view', 'comments_per_view', 'shares_per_view']
].agg(['count', 'mean', 'median'])

# Display the result
engagement_summary

likes_per_view                      \
                                        count      mean    median   
claim_status author_ban_status                                      
claim        active                      6566  0.329542  0.326538   
             banned                      1439  0.345071  0.358909   
             under review                1603  0.327997  0.320867   
opinion      active                      8817  0.219744  0.218330   
             banned                       196  0.206868  0.198483   
             under review                 463  0.226394  0.228051   

                               comments_per_view                      \
                                           count      mean    median   
claim_status author_ban_status                                         
claim        active                         6566  0.001393  0.000776   
             banned                         1439  0.001377  0.000746   
             under review                   1603  0.001367  0.000789   
opinion      active                         8817  0.000517  0.000252   
             banned                          196  0.000434  0.000193   
             under review                    463  0.000536  0.000293   

                               shares_per_view                      
                                         count      mean    median  
claim_status author_ban_status                                      
claim        active                       6566  0.065456  0.049279  
             banned                       1439  0.067893  0.051606  
             under review                 1603  0.065733  0.049967  
opinion      active                       8817  0.043729  0.032405  
             banned                        196  0.040531  0.030728  
             under review                  463  0.044472  0.035027

**Question: How does the data for claim videos and opinion videos compare or differ?**

When comparing **claim** videos and **opinion** videos across views, comments, likes, and shares, several key differences emerge:

- **Likes per view**: Claim videos tend to have a **higher like-to-view ratio** (around 0.33–0.35) compared to opinion videos (around 0.21–0.23). This suggests that viewers are more likely to like content that is framed as a claim rather than an opinion.

- **Comments per view**: Although the overall values are small, claim videos consistently receive **more comments per view** (0.00137–0.00139) than opinion videos (0.00043–0.00054), indicating greater viewer engagement and possibly more discussion or debate in response to claims.

- **Shares per view**: Claim videos also have a **higher share rate** (0.065–0.068) compared to opinion videos (0.041–0.045). This implies that claims may be perceived as more compelling or important to spread.

**Summary**: Across all metrics likes, comments, and shares **claim videos consistently show higher engagement per view than opinion videos**. This pattern could be important for modeling and content strategy, as it may reflect how users interact differently with factual or assertive content versus personal opinions.


## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.




## **PACE: Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response.

### **Given your efforts, what can you summarize for Rosie Mae Bradshaw and the TikTok data team?**

*Note for Learners: Your answer should address TikTok's request for a summary that covers the following points:*

*   What percentage of the data is comprised of claims and what percentage is comprised of opinions?
*   What factors correlate with a video's claim status?
*   What factors correlate with a video's engagement level?


### **Executive Summary for Rosie Mae Bradshaw and the TikTok Data Team**

After conducting an initial exploratory data analysis (EDA) on the TikTok dataset, the following insights have been identified:

#### **1. Distribution of Claims and Opinions**
- Out of the total 19,382 videos:
  - **Claims**: 9,608 videos (49.6%)
  - **Opinions**: 9,476 videos (48.9%)
  - **Missing claim_status**: 298 videos (1.5%)

This shows that the dataset is almost evenly split between claims and opinions, providing a balanced foundation for training a classification model.

#### **2. Correlation with Claim Status**
- **Engagement Levels**: Videos labeled as *claims* receive significantly **higher average and median view counts** compared to *opinion* videos. This suggests that claims may be more likely to go viral or attract attention.
- **Author Ban Status**: A **higher number of claim videos** are associated with accounts that are currently **banned or under review**, compared to opinion videos. This could imply that content containing claims may be more likely to violate community guidelines or trigger moderation.
- **Engagement Ratios**: Claim videos exhibit **higher likes per view, comments per view, and shares per view** than opinion videos, indicating deeper audience engagement.

#### **3. Correlation with Engagement**
- **Ban Status and Engagement**: Videos posted by **banned or under-review accounts** have **substantially higher median views, likes, and shares** than those from active accounts. This suggests that high engagement may be linked with increased scrutiny or violations.
- **Engagement Ratios**: Higher ratios of likes, comments, and shares per view are observed for videos marked as claims, especially those by banned or under-review authors. This may reflect the viral nature or controversial content of such videos.

#### **Conclusion**
Claim videos are generally associated with **higher engagement levels** and are more prevalent among banned or reviewed authors. These findings may assist in feature engineering for a machine learning model to classify future videos, and provide strategic insights into content moderation and platform safety efforts.


**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.